scrapy/utils/httpobj.py

```diff
@@ -1,10 +1,10 @@
 """Helper functions for scrapy.http objects (Request, Response)"""
 
-import weakref
 from urllib.parse import urlparse
+from weakref import WeakKeyDictionary
 
 
-_urlparse_cache = weakref.WeakKeyDictionary()
+_urlparse_cache: WeakKeyDictionary = WeakKeyDictionary()
```

Comment:
Would it be possible to type annotate keys/values here, and other similar places? It is a chance to make type annotations a bit more useful for reading (and probably using) the code; otherwise I suspect we'll never get to this - mypy error is silenced, all is good :)

Reply:
I looked into [`typing`](https://docs.python.org/3/library/typing.html) and couldn't find it, so I decided to only annotate the "outer" type. 
According to [this comment](https://bugs.python.org/issue38756#msg356318), it should be possible to annotate it as a string (`from __future__ import annotations` doesn't work in py36).
I'll update this shortly.